channels_redis/alt.py

```diff
@@ -0,0 +1,311 @@
+import asyncio
+import sys
+import traceback
+import uuid
+
+import aioredis
+import msgpack
+
+
+class RedisPubSubChannelLayer:
+    """
+    Channel Layer that uses Redis's pub/sub functionality.
+    """
+
+    def __init__(self, hosts, prefix="asgi", **kwargs):
+        assert (
+            isinstance(hosts, list) and len(hosts) > 0
+        ), "`hosts` must be a list with at least one Redis server"
+
+        self.prefix = prefix
+
+        # Each consumer gets its own *specific* channel, created with the `new_channel()` method.
+        # This dict maps `channel_name` to a queue of messages for that channel.
+        self.channels = {}
+
```

Comment:
I don't understand why this channel layer keeps a local copy of groups and the channels subscribed to groups. If this is the case, then this implementation won't scale and I think it won't be any different than the previous one.

Rather, I would suggest just let redis handle this natively and make calls from the internal client whenever we need to know which channels are subscribed.

Also, I mean that we should not keep track of clients in sets from the channel layer.
instead we need to initialize the client from the consumer class.

Reply:
You might have more than one consumer that is subscribed to a single group ("channel" in redis terminology). You can only subscribe to that channel once per redis connection. Think of it this way: Redis publishes inter-process for us (i.e. Redis does [IPC](https://en.wikipedia.org/wiki/Inter-process_communication)), and we do our own publishing intra-process (being intra-process, it's as fast as it can be...because there's no blocking or anything because we are intra-process). So the way we are doing it here is actually fastest.

Reply:
but then we have a tradeoff here.
This implementation cannot hold more than a 1000 channels as a whole, because we locally store these in sets.
In order for this implementation to be scalable we need to avoid doing so - even if it means more calls to redis.

Reply:
@acu192 
the whole point of using pub sub is to utilize the scalability redis gives. Having local caches completely destroys this purpose. Secondly, we don't have connections per consumer. We are not even using the pub sub feature properly here.
This implementation is only having two connections - one to listen and one to subscribe. In other words, it is keeping clients in a local set, then keeping track of what channels they are subscribed to, and subscribing for them via the second connection. So for a lot of consumers this won't work out - if you see what I mean. Instead, it is better to have a connection per consumer and then let redis handle all the subscriptions, natively. This would be the best implementation possible.
what do you say @carltongibson ?

Reply:
> Having local caches completely destroys this purpose.

No, local in-memory operations are ALWAYS BETTER than network I/O. Always.

The fastest channel layer you could ever have is actually the [InMemoryChannelLayer](https://github.com/django/channels/blob/588f7d3c7d0734727ffdd358d5b48145b9979258/channels/layers.py#L192). But it doesn't work if you need more than one Django server replica. Therefore, we use Redis to fill in the gaps ... communication between replicas. That's what this PR does.

Please know your voice has been heard and we have discussed this enough. Next let me discuss with the maintainers.